Error when importing CommonJS module that contains files like `module.scss.js`

### Describe the bug

I am importing a CommonJS module installed from NPM. The modue contains files with filenames like `PeoplePicker.scss.js`. These are JavaScript files, obviously, but the filename contains `scss`.

I'd expect Vite to handle these like any other JavaScript files when I run the dev server via `vite`. What actually happens is that it seems like Vite externalizes those files (forgive me if I don't fully understand what's going on, but that's my best guess), and so anything trying to import them as JavaScript is unable to do so. See the error logs.

It seems like Vite is getting tripped up by the `scss` in the filename. The same thing happens for any `${extension}.js` filename where `extension` is considered an external type. [See what I believe to be the relevant file (esbuildDepPlugin.ts).](https://github.com/vitejs/vite/blob/main/packages/vite/src/node/optimizer/esbuildDepPlugin.ts)

### Reproduction

https://github.com/noahmpauls/vite-commonjs-external-bug

### Steps to reproduce

Run `npm install` followed by `npm run -w app dev`. The reproduction isn't applicable to the `build` script.

### System Info

```shell
System:
    OS: Linux 6.7 Arch Linux
    CPU: (8) x64 Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
    Memory: 2.56 GB / 7.42 GB
    Container: Yes
    Shell: 5.2.26 - /usr/bin/bash
  Binaries:
    Node: 20.9.0 - ~/.local/share/volta/tools/image/node/20.9.0/bin/node
    npm: 10.2.5 - ~/.local/share/volta/tools/image/npm/10.2.5/bin/npm
```


### Used Package Manager

npm

### Logs

<details>
<summary>Click to expand!</summary>

```shell
  VITE v5.1.5  ready in 150 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
  ➜  press h + enter to show help
✘ [ERROR] Could not resolve "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.tsx.js"

    vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.tsx.js:1:24:
      1 │ ... { default } from "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.tsx.js"...
        ╵                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  The plugin "vite:dep-pre-bundle" didn't set a resolve directory for the file
  "vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.tsx.js",
  so esbuild did not search for
  "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.tsx.js" on the file
  system.

✘ [ERROR] Could not resolve "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.stylus.js"

    vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.stylus.js:1:24:
      1 │ ... { default } from "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.stylus....
        ╵                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  The plugin "vite:dep-pre-bundle" didn't set a resolve directory for the file
  "vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.stylus.js",
  so esbuild did not search for
  "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.stylus.js" on the
  file system.

✘ [ERROR] Could not resolve "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.astro.js"

    vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.astro.js:1:24:
      1 │ ... { default } from "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.astro.j...
        ╵                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  The plugin "vite:dep-pre-bundle" didn't set a resolve directory for the file
  "vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.astro.js",
  so esbuild did not search for
  "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.astro.js" on the file
  system.

✘ [ERROR] Could not resolve "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.scss.js"

    vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.scss.js:1:24:
      1 │ ... { default } from "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.scss.js...
        ╵                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  The plugin "vite:dep-pre-bundle" didn't set a resolve directory for the file
  "vite:dep-pre-bundle:external-conversion:/home/noahpauls/src/vite-scss-bug/package/test.scss.js",
  so esbuild did not search for
  "vite-dep-pre-bundle-external:/home/noahpauls/src/vite-scss-bug/package/test.scss.js" on the file
  system.
```
</details>

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/optimizer/esbuildDepPlugin.ts
+++ b/packages/vite/src/node/optimizer/esbuildDepPlugin.ts
@@ -155,6 +155,14 @@ export function esbuildDepPlugin(
           const resolved = await resolve(id, importer, kind)
           if (resolved) {
             if (kind === 'require-call') {
+              // #16116 fix: Import the module.scss path, which is actually module.scss.js
+              if (resolved.endsWith('.js')) {
+                return {
+                  path: resolved,
+                  external: false,
+                }
+              }
+
               // here it is not set to `external: true` to convert `require` to `import`
               return {
                 path: resolved,

--- a/playground/optimize-deps/__tests__/optimize-deps.spec.ts
+++ b/playground/optimize-deps/__tests__/optimize-deps.spec.ts
@@ -331,3 +331,18 @@ test.runIf(isServe)('warn on incompatible dependency', () => {
     ),
   )
 })
+
+test('import the CommonJS external package that omits the js suffix', async () => {
+  await expectWithRetry(() => page.textContent('.external-package-js')).toBe(
+    'okay',
+  )
+  await expectWithRetry(() =>
+    page.textContent('.external-package-scss-js'),
+  ).toBe('scss')
+  await expectWithRetry(() =>
+    page.textContent('.external-package-astro-js'),
+  ).toBe('astro')
+  await expectWithRetry(() =>
+    page.textContent('.external-package-tsx-js'),
+  ).toBe('tsx')
+})

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/index.js
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/index.js
@@ -1,0 +1,6 @@
+const { okay } = require('./test.okay')
+const { scss } = require('./test.scss')
+const { astro } = require('./test.astro')
+const { tsx } = require('./test.tsx')
+
+module.exports = { okay, scss, astro, tsx }

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/package.json
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/package.json
@@ -1,0 +1,5 @@
+{
+  "name": "@vitejs/test-dep-cjs-external-package-omit-js-suffix",
+  "version": "0.0.1",
+  "main": "index.js"
+}

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/package.json
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/package.json
@@ -1,5 +1,6 @@
 {
   "name": "@vitejs/test-dep-cjs-external-package-omit-js-suffix",
-  "version": "0.0.1",
+  "private": true,
+  "version": "0.0.0",
   "main": "index.js"
 }

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.astro.js
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.astro.js
@@ -1,0 +1,5 @@
+function astro() {
+  return 'astro'
+}
+
+module.exports = { astro }

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.okay.js
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.okay.js
@@ -1,0 +1,5 @@
+function okay() {
+  return 'okay'
+}
+
+module.exports = { okay }

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.scss.js
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.scss.js
@@ -1,0 +1,5 @@
+function scss() {
+  return 'scss'
+}
+
+module.exports = { scss }

--- a/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.tsx.js
+++ b/playground/optimize-deps/dep-cjs-external-package-omit-js-suffix/test.tsx.js
@@ -1,0 +1,5 @@
+function tsx() {
+  return 'tsx'
+}
+
+module.exports = { tsx }

--- a/playground/optimize-deps/index.html
+++ b/playground/optimize-deps/index.html
@@ -110,6 +110,25 @@
 
 <script type="module" src="./long-file-name.js"></script>
 
+<h2>Import the CommonJS external package that omits the js suffix</h2>
+<div class="external-package-js"></div>
+<div class="external-package-scss-js"></div>
+<div class="external-package-astro-js"></div>
+<div class="external-package-tsx-js"></div>
+<script type="module">
+  import {
+    astro,
+    okay,
+    scss,
+    tsx,
+  } from '@vitejs/test-dep-cjs-external-package-omit-js-suffix'
+  console.log(astro(), 'astroastroastroastro')
+  text('.external-package-js', okay())
+  text('.external-package-scss-js', scss())
+  text('.external-package-astro-js', astro())
+  text('.external-package-tsx-js', tsx())
+</script>
+
 <script>
   function text(el, text) {
     document.querySelector(el).textContent = text

--- a/playground/optimize-deps/index.html
+++ b/playground/optimize-deps/index.html
@@ -122,7 +122,7 @@
     scss,
     tsx,
   } from '@vitejs/test-dep-cjs-external-package-omit-js-suffix'
-  console.log(astro(), 'astroastroastroastro')
+
   text('.external-package-js', okay())
   text('.external-package-scss-js', scss())
   text('.external-package-astro-js', astro())

--- a/playground/optimize-deps/package.json
+++ b/playground/optimize-deps/package.json
@@ -36,6 +36,7 @@
     "@vitejs/test-dep-with-optional-peer-dep-submodule": "file:./dep-with-optional-peer-dep-submodule",
     "@vitejs/test-dep-non-optimized": "file:./dep-non-optimized",
     "@vitejs/test-added-in-entries": "file:./added-in-entries",
+    "@vitejs/test-dep-cjs-external-package-omit-js-suffix": "file:./dep-cjs-external-package-omit-js-suffix",
     "lodash-es": "^4.17.21",
     "@vitejs/test-nested-exclude": "file:./nested-exclude",
     "phoenix": "^1.7.11",

--- a/playground/optimize-deps/vite.config.js
+++ b/playground/optimize-deps/vite.config.js
@@ -17,6 +17,7 @@ export default defineConfig({
     include: [
       '@vitejs/test-dep-linked-include',
       '@vitejs/test-nested-exclude > @vitejs/test-nested-include',
+      '@vitejs/test-dep-cjs-external-package-omit-js-suffix',
       // will throw if optimized (should log warning instead)
       '@vitejs/test-non-optimizable-include',
       '@vitejs/test-dep-optimize-exports-with-glob/**/*',
@@ -52,6 +53,9 @@ export default defineConfig({
         if (msg.message.includes('Buffer')) return
         warn(msg)
       },
+    },
+    commonjsOptions: {
+      include: ['./dep-cjs-external-package-omit-js-suffix'],
     },
   },
 

--- a/playground/optimize-deps/vite.config.js
+++ b/playground/optimize-deps/vite.config.js
@@ -54,9 +54,6 @@ export default defineConfig({
         warn(msg)
       },
     },
-    commonjsOptions: {
-      include: ['./dep-cjs-external-package-omit-js-suffix'],
-    },
   },
 
   plugins: [

